### PR TITLE
[FIX] MSGF+ enzymes

### PR DESCRIPTION
**Pull request by jpfeuffer**
Unifies the enzymes in MSGF+ Adapter to the ones from our EnzymesDB.
Added some new Enzymes that do not have a PSI obo ID yet. Hope this is fine.
Adresses #2355

### CODE REVIEW

File: src/topp/MSGFPlusAdapter.cpp [hunk 1]
```diff
@@ -474,7 +477,8 @@ class MSGFPlusAdapter :
     // no need to handle "not found" case - would have given error during parameter parsing:
     Int fragment_method_code = ListUtils::getIndex<String>(fragment_methods_, getStringOption_("fragment_method"));
     Int instrument_code = ListUtils::getIndex<String>(instruments_, getStringOption_("instrument"));
-    Int enzyme_code = ListUtils::getIndex<String>(enzymes_, enzyme);
+    Int enzyme_code = EnzymesDB::getInstance()->getEnzyme(enzyme)->getMSGFid();
+    std::cout << enzyme_code << std::endl;
```
Comment (timosachsenberg) [inline src/topp/MSGFPlusAdapter.cpp]: debug code?

File: src/topp/MSGFPlusAdapter.cpp [hunk 2]
```diff
@@ -464,6 +466,7 @@ class MSGFPlusAdapter :
 
     // parameters also used by OpenMS (see idXML creation below):
     String enzyme = getStringOption_("enzyme");
+    std::cout << enzyme << std::endl;
```
Comment (timosachsenberg) [inline src/topp/MSGFPlusAdapter.cpp]: debug code?

File: src/topp/MSGFPlusAdapter.cpp (hunk 1 above)
Comment (jpfeuffer) [inline src/topp/MSGFPlusAdapter.cpp]: Ah thx!

File: src/topp/MSGFPlusAdapter.cpp [hunk 3]
```diff
@@ -162,7 +164,7 @@ class MSGFPlusAdapter :
     registerStringOption_("instrument", "<choice>", instruments_[0], "Instrument that generated the data ('low_res'/'high_res' refer to LCQ and LTQ instruments; MS-GF+ parameter '-inst')", false);
     setValidStrings_("instrument", instruments_);
 
-    registerStringOption_("enzyme", "<choice>", enzymes_[1], "Enzyme used for digestion, or type of cleavage (MS-GF+ parameter '-e')", false);
+    registerStringOption_("enzyme", "<choice>", enzymes_[6], "Enzyme used for digestion, or type of cleavage (MS-GF+ parameter '-e')", false);
```
Comment (cbielow) [inline src/topp/MSGFPlusAdapter.cpp]: nice PR!
Could you add a comment here like:
"Note: msgf+ only supports '/P' cutting rules."

File: share/OpenMS/CHEMISTRY/Enzymes.xml [hunk 4]
```diff
@@ -59,6 +78,15 @@
...
       <ITEM name="CometID" value="10" type="int" />
     </NODE>
+    <NODE name="Chymotrypsin/P">
+      <ITEM name="Name" value="Chymotrypsin/P" type="string" />
+      <ITEM name="RegEx" value="(?&lt;=[FLWY])" type="string" />
+      <ITEM name="RegExDescription" value="Chymotrypsin cleaves following F, Y, W or L residue." type="string" />
+      <ITEM name="NTermGain" value="H" type="string" />
+      <ITEM name="CTermGain" value="OH" type="string" />
+      <ITEM name="XTandemID" value="[FYWL]|[X]" type="string" />
+      <ITEM name="MSGFID" value="[FYWL]|[X]" type="string" />
```
Comment (cbielow) [inline share/OpenMS/CHEMISTRY/Enzymes.xml]: should be value=2

File: share/OpenMS/CHEMISTRY/Enzymes.xml [hunk 5]
```diff
@@ -188,11 +219,21 @@
       <ITEM name="XTandemID" value="[DE]|[X]" type="string" />
       <ITEM name="OMSSAID" value="20" type="int" />
       <ITEM name="CometID" value="8" type="int" />
+      <ITEM name="MSGFID" value="1" type="int" />
```
Comment (cbielow) [inline share/OpenMS/CHEMISTRY/Enzymes.xml]: should be 5?

File: share/OpenMS/CHEMISTRY/Enzymes.xml [hunk 6]
```diff
@@ -37,6 +46,16 @@
...
+      <ITEM name="MSGFID" value="7" type="int" />
+    </NODE>
+    <NODE name="Asp-N/B">
+      <ITEM name="Name" value="Asp-N/B" type="string" />
+      <ITEM name="RegEx" value="(?=[D])" type="string" />
+      <ITEM name="RegExDescription" value="Asp-N/B cleaves before D." type="string" />
+      <ITEM name="NTermGain" value="OH" type="string" />
+      <ITEM name="CTermGain" value="H" type="string" />
+      <ITEM name="XTandemID" value="[X]|[D]" type="string" />
+      <ITEM name="MSGFID" value="7" type="int" />
```
Comment (cbielow) [inline share/OpenMS/CHEMISTRY/Enzymes.xml]: id=7 shows up twice. Which one is the right one?

**Comment by jpfeuffer** [comment]
Thanks for the comments @cbielow maybe the merge messed up some things. I will double check the numbers and a comment.

File: src/openms/source/CHEMISTRY/Enzyme.cpp [hunk 7]
```diff
@@ -220,6 +225,16 @@ namespace OpenMS
...
   }
 
+  void Enzyme::setMSGFID(Int value)
+  {
+    msgf_id_ = value;
+  }
+
+  Int Enzyme::getMSGFID() const
+  {
+      return msgf_id_;
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: indentation

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp [hunk 8]
```diff
@@ -307,4 +315,16 @@ namespace OpenMS
       }
     }
   }
+
+  void EnzymesDB::getAllMSGFNames(vector<String>& all_names) const
+  {
+    all_names.clear();
+    for (ConstEnzymeIterator it = const_enzymes_.begin(); it != const_enzymes_.end(); ++it)
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: consider using something like: 
for (const Enzyme& e ; const_enzymes_)
{ ... e.getMSGFId() ...) 
for better readability in future PRs

**Comment by timosachsenberg** [comment]
pxd need to be adapted, too

File: share/OpenMS/CHEMISTRY/Enzymes.xml [hunk 9]
```diff
@@ -95,11 +122,12 @@
       <ITEM name="Name" value="Lys-N" type="string" />
       <ITEM name="RegEx" value="(?=K)" type="string" />
       <ITEM name="RegExDescription" value="Lys-N cuts before K" type="string" />
-      <ITEM name="NTermGain" value="H" type="string" />
-      <ITEM name="CTermGain" value="OH" type="string" />
+      <ITEM name="NTermGain" value="OH" type="string" />
```
Comment (timosachsenberg) [inline share/OpenMS/CHEMISTRY/Enzymes.xml]: why is this changed again?

File: share/OpenMS/CHEMISTRY/Enzymes.xml [hunk 10]
```diff
@@ -38,6 +47,15 @@
       <ITEM name="OMSSAID" value="12" type="int" />
       <ITEM name="CometID" value="6" type="int" />
     </NODE>
+    <NODE name="Asp-N/B">
+      <ITEM name="Name" value="Asp-N/B" type="string" />
+      <ITEM name="RegEx" value="(?=[D])" type="string" />
+      <ITEM name="RegExDescription" value="Asp-N/B cleaves before D." type="string" />
+      <ITEM name="NTermGain" value="OH" type="string" />
```
Comment (timosachsenberg) [inline share/OpenMS/CHEMISTRY/Enzymes.xml]: sure this is NTermGain OH?

**Comment by jpfeuffer** [comment]
Wait for fix of Terminal gains please

File: share/OpenMS/CHEMISTRY/Enzymes.xml (hunk 6 above)
Comment (jpfeuffer) [inline share/OpenMS/CHEMISTRY/Enzymes.xml]: This one.

**Comment by jpfeuffer** [comment]
Ok I looked at the terminal gains. There seem to be no irregularities except for CNBr (and there can be two different types of modifications to the cleaved Met depending on pH). So either we remove the support for cleavage modification (which is not used anywhere anyway) or we need to introduce multiple possible gains/losses.
Otherwise, ready to merge.

**Comment by jpfeuffer** [comment]
With some additional computational effort, this can be done like suggested by Mascot, with variable modifications (e.g. Met->Hsn)

**Comment by jpfeuffer** [comment]
Seems like it is not even possible to define a mix of losses and gains. 
```
exception of type Parse Error occured in line 168, function void OpenMS::EnzymesDB::readEnzymesFromFile_(const OpenMS::String&) of /home/travis/build/OpenMS/OpenMS/src/openms/source/CHEMISTRY/EnzymesDB.cpp
error message:  in: Could not convert string '-' to an integer value
```
Why does our reader try to parse the string "-SCH3 +O" as an integer?
Well I'll just remove.

**Comment by timosachsenberg** [comment]
+1 and we might think about removing the gain entries in the future
